### PR TITLE
Add support for recording iOS screens

### LINKED ISSUE #25: Support recording iOS screen
[Issuehunt badges](https://issuehunt.io/r/wulkano/Aperture/issues/25)

I've started the work in https://github.com/wulkano/aperture/commit/ce2e2289bf0ca92aeb6bc4337c8d4826a09d0642.

We still need to add support for recording the iOS screen. It's not considered a screen, but rather a normal video input, so we need to add some custom logic for that.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $100.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($100.00)
### Submitted pull Requests
- [#69 Add ios screen recording feature](https://issuehunt.io/r/wulkano/Aperture/pull/69)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/wulkano/Aperture/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by Ybrin**
Fixes #25 

I added a second initializer which accepts an `AVCaptureDevice` as the `iosDevice` and records the iOS screen together with an optional given audio device.

I made the audio device default to nil for recording iOS screens as I don't think it's a good default to record audio with the system default if you want to record an iOS screen. Callers can still set it to whatever they want.

I also changed the `Devices.ios()` function a little bit. It should now work like that.

The following is an example snippet which records 5 seconds of some connected iOS screen (the first one it finds) without audio.

```Swift
let url = URL(fileURLWithPath: "/Users/tester/Downloads/ios_screen.mp4")

let aperture = try! Aperture(destination: url, iosDevice: AVCaptureDevice(uniqueID: Devices.ios()[0]["id"]!)!)

aperture.start()

sleep(5)

aperture.stop()
```

@sindresorhus @skllcrn Could you review this for me?

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#25: Support recording iOS screen](https://issuehunt.io/repos/65488992/issues/25)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by Ybrin** [comment]
@sindresorhus Will be done.

Sure, I can open a PR to aperture-node once this one is merged and tagged so I can use it there.

File: Sources/Aperture/Aperture.swift [hunk 1]
```diff
@@ -92,6 +92,59 @@ public final class Aperture: NSObject {
...
+
+    if session.canAddOutput(output) {
+      session.addOutput(output)
+    } else {
+      throw ApertureError.couldNotAddOutput
+    }
+
+    if let videoCodec = videoCodec {
+      output.setOutputSettings([AVVideoCodecKey: videoCodec], for: output.connection(with: .video)!)
+    }
```
[the remark this answers is not in the document]
Comment (Ybrin) [reply Sources/Aperture/Aperture.swift]: Done

File: Sources/Aperture/Aperture.swift [hunk 2]
```diff
@@ -92,6 +92,59 @@ public final class Aperture: NSObject {
     super.init()
   }
 
+  public init(
```
[the remark this answers is not in the document]
Comment (Ybrin) [reply Sources/Aperture/Aperture.swift]: Done

**Comment by Ybrin** [comment]
@sindresorhus I think I'm done with this. Could you give it a rough review and make a release so I can start working on aperture-node? Thanks.

**Comment by Ybrin** [comment]
@sindresorhus @skllcrn Is there anything left to do for this PR to be approved? I would like to implement this in aperture-node once it's merged...

**Comment by Ybrin** [comment]
@matheuss Maybe? This feature is really done from my perspective...

**Comment by matheuss** [comment]
@Ybrin very long ago (almost 3y at this point) since I last touched Swift, so I'm not able to help much on reviewing 😓

**Comment by sindresorhus** [comment]
> The following is an example snippet which records 5 seconds of some connected iOS screen (the first one it finds) without audio.

Can you add this to the project so we can just run some command to test it?

**Comment by Ybrin** [comment]
@sindresorhus What do you mean? Do you want me to create an example project in this repository or create a new repository as a sample?

**Comment by sindresorhus** [comment]
In this project. Doesn't have to be an example project. Just an single command (Like `$ ./example` which runs `swiftc` and something internally) to be able to run the above snippet and get the example recorded iOS video in the same directory.

**Comment by Ybrin** [comment]
@sindresorhus I added a shellscript which runs a simple example to record 5 seconds of the first found iOS device. Start it by running `./screen_recording_sample.sh` in the project directory

**Comment by Ybrin** [comment]
Btw, the example only works if there is an iOS device connected, obviously...

@sindresorhus Once this is merged I can implement it in aperture-node as you told me before.

**Comment by sindresorhus** [comment]
Released: https://github.com/wulkano/Aperture/releases/tag/v0.3.0

Note that I did some other breaking changes.

**Comment by sindresorhus** [comment]
> Sure, I can open a PR to aperture-node once this one is merged and tagged so I can use it there.

@Ybrin Still interested in finishing that? ^

**Comment by Ybrin** [comment]
@sindresorhus Yes sure, I can do this in aperture-node.

Btw, do you know why I was not rewarded on Issuehunt for this? The issue is closed and the PR merged...

https://issuehunt.io/r/wulkano/Aperture/issues/25

Seems strange to me but the support doesn't answer...

**Comment by sindresorhus** [comment]
Sorry. That’s my fault. IssueHunt doesn’t notify me when an issue with a bounty is closed and I have to manually reward. Done now.

**Pull request merged**